error: cannot take the address of an rvalue of type 'FILE *' when building on OpenBSD

On [line 137 of log.c](https://github.com/firehol/netdata/blob/master/src/log.c#L137) (and a few lines after that), netdata tries to get the address of `stdout` and co., which is a compilation error on OpenBSD (see issue title), and is [not technically within C specifications](http://openwall.com/lists/musl/2018/02/02/2).

I have some experience developing in C and could probably put a patch together which uses freopen(3), as suggested by the musl developer in the mailing list link above. Should I investigate patching it myself, or is this a trivial enough change that you'd rather just do it yourself?

Well, it turns out it does not work on a few versions of gcc/Linux:

```
log.c: In function 'reopen_all_log_files':
log.c:142:16: error: assignment of read-only variable 'stdout'
         stdout = open_log_file(STDOUT_FILENO, stdout, stdout_filename, &output_log_syslog, 0, NULL);
                ^
log.c:145:16: error: assignment of read-only variable 'stderr'
         stderr = open_log_file(STDERR_FILENO, stderr, stderr_filename, &error_log_syslog, 0, NULL);
                ^
log.c: In function 'open_all_log_files':
log.c:153:11: error: assignment of read-only variable 'stdin'
     stdin = open_log_file(STDIN_FILENO, stdin, "/dev/null", NULL, 0, NULL);
           ^
log.c:155:12: error: assignment of read-only variable 'stdout'
     stdout = open_log_file(STDOUT_FILENO, stdout, stdout_filename, &output_log_syslog, 0, NULL);
            ^
log.c:156:12: error: assignment of read-only variable 'stderr'
     stderr = open_log_file(STDERR_FILENO, stderr, stderr_filename, &error_log_syslog, 0, NULL);
            ^
```

So, even the `freopen()` method is not valid.

--- a/src/log.c
+++ b/src/log.c
@@ -99,15 +99,15 @@ static FILE *open_log_file(int fd, FILE *fp, const char *filename, int *enabled_
         }
     }
 
-    if(devnull && is_stdaccess) {
-        fd = -1;
-        fp = NULL;
-    }
-
     // if there is a level-2 file pointer
     // flush it before switching the level-1 fds
     if(fp)
         fflush(fp);
+
+    if(devnull && is_stdaccess) {
+        fd = -1;
+        fp = NULL;
+    }
 
     if(fd != f && fd != -1) {
         // it automatically closes
@@ -139,10 +139,10 @@ static FILE *open_log_file(int fd, FILE *fp, const char *filename, int *enabled_
 
 void reopen_all_log_files() {
     if(stdout_filename)
-        stdout = open_log_file(STDOUT_FILENO, stdout, stdout_filename, &output_log_syslog, 0, NULL);
+        open_log_file(STDOUT_FILENO, stdout, stdout_filename, &output_log_syslog, 0, NULL);
 
     if(stderr_filename)
-        stderr = open_log_file(STDERR_FILENO, stderr, stderr_filename, &error_log_syslog, 0, NULL);
+        open_log_file(STDERR_FILENO, stderr, stderr_filename, &error_log_syslog, 0, NULL);
 
     if(stdaccess_filename)
          stdaccess = open_log_file(stdaccess_fd, stdaccess, stdaccess_filename, &access_log_syslog, 1, &stdaccess_fd);
@@ -150,10 +150,10 @@ void reopen_all_log_files() {
 
 void open_all_log_files() {
     // disable stdin
-    stdin = open_log_file(STDIN_FILENO, stdin, "/dev/null", NULL, 0, NULL);
+    open_log_file(STDIN_FILENO, stdin, "/dev/null", NULL, 0, NULL);
 
-    stdout = open_log_file(STDOUT_FILENO, stdout, stdout_filename, &output_log_syslog, 0, NULL);
-    stderr = open_log_file(STDERR_FILENO, stderr, stderr_filename, &error_log_syslog, 0, NULL);
+    open_log_file(STDOUT_FILENO, stdout, stdout_filename, &output_log_syslog, 0, NULL);
+    open_log_file(STDERR_FILENO, stderr, stderr_filename, &error_log_syslog, 0, NULL);
     stdaccess = open_log_file(stdaccess_fd, stdaccess, stdaccess_filename, &access_log_syslog, 1, &stdaccess_fd);
 }
 

--- a/src/log.c
+++ b/src/log.c
@@ -57,7 +57,7 @@ static inline void log_unlock() {
     netdata_mutex_unlock(&log_mutex);
 }
 
-int open_log_file(int fd, FILE **fp, const char *filename, int *enabled_syslog) {
+static FILE *open_log_file(int fd, FILE *fp, const char *filename, int *enabled_syslog, int is_stdaccess, int *fd_ptr) {
     int f, devnull = 0;
 
     if(!filename || !*filename || !strcmp(filename, "none") ||  !strcmp(filename, "/dev/null")) {
@@ -76,8 +76,10 @@ int open_log_file(int fd, FILE **fp, const char *filename, int *enabled_syslog) 
     // don't do anything if the user is willing
     // to have the standard one
     if(!strcmp(filename, "system")) {
-        if(fd != -1 && fp != &stdaccess)
-            return fd;
+        if(fd != -1 && !is_stdaccess) {
+            if(fd_ptr) *fd_ptr = fd;
+            return fp;
+        }
 
         filename = "stderr";
     }
@@ -92,19 +94,20 @@ int open_log_file(int fd, FILE **fp, const char *filename, int *enabled_syslog) 
         f = open(filename, O_WRONLY | O_APPEND | O_CREAT, 0664);
         if(f == -1) {
             error("Cannot open file '%s'. Leaving %d to its default.", filename, fd);
-            return fd;
+            if(fd_ptr) *fd_ptr = fd;
+            return fp;
         }
     }
 
-    if(devnull && fp == &stdaccess) {
+    if(devnull && is_stdaccess) {
         fd = -1;
-        *fp = NULL;
+        fp = NULL;
     }
 
     // if there is a level-2 file pointer
     // flush it before switching the level-1 fds
-    if(fp && *fp)
-        fflush(*fp);
+    if(fp)
+        fflush(fp);
 
     if(fd != f && fd != -1) {
         // it automatically closes
@@ -112,44 +115,46 @@ int open_log_file(int fd, FILE **fp, const char *filename, int *enabled_syslog) 
         if (t == -1) {
             error("Cannot dup2() new fd %d to old fd %d for '%s'", f, fd, filename);
             close(f);
-            return fd;
+            if(fd_ptr) *fd_ptr = fd;
+            return fp;
         }
         // info("dup2() new fd %d to old fd %d for '%s'", f, fd, filename);
         close(f);
     }
     else fd = f;
 
-    if(fp && !*fp) {
-        *fp = fdopen(fd, "a");
-        if (!*fp)
+    if(!fp) {
+        fp = fdopen(fd, "a");
+        if (!fp)
             error("Cannot fdopen() fd %d ('%s')", fd, filename);
         else {
-            if (setvbuf(*fp, NULL, _IOLBF, 0) != 0)
+            if (setvbuf(fp, NULL, _IOLBF, 0) != 0)
                 error("Cannot set line buffering on fd %d ('%s')", fd, filename);
         }
     }
 
-    return fd;
+    if(fd_ptr) *fd_ptr = fd;
+    return fp;
 }
 
 void reopen_all_log_files() {
     if(stdout_filename)
-        open_log_file(STDOUT_FILENO, (FILE **)&stdout, stdout_filename, &output_log_syslog);
+        stdout = open_log_file(STDOUT_FILENO, stdout, stdout_filename, &output_log_syslog, 0, NULL);
 
     if(stderr_filename)
-        open_log_file(STDERR_FILENO, (FILE **)&stderr, stderr_filename, &error_log_syslog);
+        stderr = open_log_file(STDERR_FILENO, stderr, stderr_filename, &error_log_syslog, 0, NULL);
 
     if(stdaccess_filename)
-        stdaccess_fd = open_log_file(stdaccess_fd, (FILE **)&stdaccess, stdaccess_filename, &access_log_syslog);
+         stdaccess = open_log_file(stdaccess_fd, stdaccess, stdaccess_filename, &access_log_syslog, 1, &stdaccess_fd);
 }
 
 void open_all_log_files() {
     // disable stdin
-    open_log_file(STDIN_FILENO, (FILE **)&stdin, "/dev/null", NULL);
+    stdin = open_log_file(STDIN_FILENO, stdin, "/dev/null", NULL, 0, NULL);
 
-    open_log_file(STDOUT_FILENO, (FILE **)&stdout, stdout_filename, &output_log_syslog);
-    open_log_file(STDERR_FILENO, (FILE **)&stderr, stderr_filename, &error_log_syslog);
-    stdaccess_fd = open_log_file(stdaccess_fd, (FILE **)&stdaccess, stdaccess_filename, &access_log_syslog);
+    stdout = open_log_file(STDOUT_FILENO, stdout, stdout_filename, &output_log_syslog, 0, NULL);
+    stderr = open_log_file(STDERR_FILENO, stderr, stderr_filename, &error_log_syslog, 0, NULL);
+    stdaccess = open_log_file(stdaccess_fd, stdaccess, stdaccess_filename, &access_log_syslog, 1, &stdaccess_fd);
 }
 
 // ----------------------------------------------------------------------------

